jupyterhub/app.py

```diff
@@ -1975,25 +1974,42 @@ async def init_groups(self):
                 group.users.append(user)
         db.commit()
 
-    async def init_roles(self):
+    async def init_role_creation(self):
         """Load default and predefined roles into the database"""
-        db = self.db
-        # tokens are added separately
-        role_bearers = ['users', 'services', 'groups']
-
-        # load default roles
         self.log.debug('Loading default roles to database')
         default_roles = roles.get_default_roles()
-        for role in default_roles:
-            roles.create_role(db, role)
+        default_role_names = [r['name'] for r in default_roles]
+        init_roles = default_roles + self.load_roles
+        if not orm.Role.find(self.db, name='admin'):
+            self._rbac_upgrade = True
+        for role in self.db.query(orm.Role):
+            if role.name not in default_role_names:
```

Comment:
This will delete all roles apart from default roles from the database on every startup and then recreate the ones present in the config file again (in the code below). Shouldn't we only delete the roles not present in the config anymore ?

Then again, deleting the roles automatically strips the roles from all previous bearers so we can start with clean plate, reassigning from scratch. Is that the intention?

Reply:
The latter was indeed my intention, the reason being that I don't see a way to (efficiently) check if a defined role was previously differently defined in the configuration file. Since this actually also holds for the default roles, I went on to delete all roles prior to creating them

Reply:
The `_overwrite_role()` called within `create_role()` in `roles.py` is checking if the 'description' and 'scopes' keys are different in the config than what is stored in db and overwrites them if yes. Maybe something like this? But starting from a clean plate completely reduces the chance of mismatch anywhere along the way so I am actually more inclined towards your solution here.

Reply:
Great, then I'll start from an empty role db on startup, thanks!